### ISSUE TITLE
Deadlock in decoder while seeking

### ISSUE THREAD

**Issue by dougnazar** [Freescale/gstreamer-imx#111]
I've been dealing with a rare deadlock while seeking. 

The decoding thread is holding the pad stream lock while blocking in **gst_imx_vpu_decoder_context_wait_until_decoding_possible()**. The seek then tries to flush the queue which also takes the pad stream lock. We're no longer processing frames so the decoding thread never releases.

Attached is a debugging patch I've been testing. It's a horrible hack but it does fix the issue. I'm sure that I'm missing something and there is an easier/better way to fix this but I'm still learning how all the pieces fit together.

[backtrace-multiqueue.txt](https://github.com/Freescale/gstreamer-imx/files/429170/backtrace-multiqueue.txt)
[backtrace-seek.txt](https://github.com/Freescale/gstreamer-imx/files/429171/backtrace-seek.txt)
[gst-plugins-imx-debug.txt](https://github.com/Freescale/gstreamer-imx/files/429172/gst-plugins-imx-debug.txt)

**Comment by dv1** [comment]
Hmm. This is indeed a concern. But one simple solution to this could be to just release the stream lock before and re-lock it after that function call. The activity inside that function should not require the stream lock. I'll ask if there is a problem with this approach.

**Comment by dougnazar** [comment]
Can we access it? I believe it's the upstream src pad stream lock on the multiqueue created by decodebin.

**Comment by dv1** [comment]
Right.. the stream lock might be somewhere else. However, I might have a solution, but unfortunately do not currently have the time to test it. Can you see if it works?

[0001-vpu-decoder-release-decoder-context-condition-during.patch.zip](https://github.com/Freescale/gstreamer-imx/files/434125/0001-vpu-decoder-release-decoder-context-condition-during.patch.zip)

The idea is that the stream lock is only held during the FLUSH_STOP event, not FLUSH_START. So, between FLUSH_START and FLUSH_STOP, the condition variable must not get invoked. This is done by calling gst_imx_vpu_decoder_context_set_no_wait() and setting no_wait to TRUE until flushing finishes.

**Comment by dougnazar** [comment]
This seems like a much better approach. I saw the no_wait flag, just didn't know where to hook it up to the flush. I've had it running my stress test for 12h and no lockup.

Needed to add the following to the top of the function:
**GstImxVpuDecoder *vpu_decoder = GST_IMX_VPU_DECODER(decoder);**

s/element/vpu_decoder/

and the second message should be **Disabling**.

Otherwise, looks good. Thanks for your help.

**Comment by dv1** [comment]
I pushed this fix now. Does it work for you?

**Comment by dougnazar** [comment]
I was actually going to message you tomorrow or Monday. I ran it for a day or so in my stress test without issues (usually fail in an hour or so). I pushed it to production on the 28th and have had no issues so far (usually several rooms would hang on Friday & Saturday).

So it looks fixed as far as I can tell. Thanks.

The only thing I noticed in the patch you committed is line 477 should be **Disabling no_wait mode...**
